From 69186ac7e0f398f0007a86ba59f9733ec1582902 Mon Sep 17 00:00:00 2001
From: Ariel Aviad <aaviad@habana.ai>
Date: Thu, 15 May 2025 09:36:53 +0300
Subject: [PATCH] [SW-221488] boot info: prevent integer overflow

Casts descriptor sizes to 64-bit in `boot_info_has_enough_space`
to prevent potential integer overflows during size calculations,
ensuring accurate space reporting.

Change-Id: I3973184ff9bef48f7a9d0294751bf7643542d2b7
Signed-off-by: Ariel Aviad <aaviad@habana.ai>
---

diff --git a/boot_info.c b/boot_info.c
index 5a911d7..f875167 100644
--- a/boot_info.c
+++ b/boot_info.c
@@ -579,7 +579,9 @@
 
 static bool boot_info_has_enough_space(enum hl_boot_info_type type, uint32_t data_size)
 {
-	uint32_t i, num_descs, ptrs_sz, total_sz = 0;
+	uint64_t total_sz = 0;
+	size_t ptrs_sz = 0;
+	uint32_t i, num_descs;
 	struct hl_boot_info_desc *desc;
 	struct info_set_list **info;
 	struct hl_boot_info_ptrs *ptrs;
@@ -598,7 +600,7 @@
 		if (*info)
 			continue;
 
-		total_sz += ROUND_UP(desc->size, 4);
+		total_sz += ROUND_UP((uint64_t)desc->size, 4);
 		descs_count++;
 
 		if (type == desc->type)
@@ -616,15 +618,17 @@
 		if (!desc)
 			descs_count++;
 
-		total_sz += ROUND_UP((*info)->size, 4);
+		total_sz += ROUND_UP((uint64_t)(*info)->size, 4);
 	}
 
 	if (!is_desc_exists)
 		descs_count++;
 
-	total_sz += ptrs_sz + (ptrs->desc_sz + sizeof(uint32_t)) * descs_count + data_size;
+	total_sz += ptrs_sz + (uint64_t)(ptrs->desc_sz + sizeof(uint32_t)) * descs_count +
+		    (uint64_t)data_size;
+	LOG_DBG("Total size: %llu", total_sz);
 
-	return total_sz > CONFIG_BOOTINFO_SIZE;
+	return (total_sz > CONFIG_BOOTINFO_SIZE);
 }
 
 static void hl_boot_info_sort_in_place(struct info_set_list **info)
